lib/matplotlib/widgets.py

```diff
@@ -2007,113 +2050,216 @@ def new_axes(self, ax):
                               **self.rectprops)
-        if self.span_stays:
-            self.stay_rect = Rectangle((0, 0), w, h,
-                                       transform=trans,
-                                       visible=False,
-                                       **self.rectprops)
-            self.stay_rect.set_animated(False)
-            self.ax.add_patch(self.stay_rect)
 
         self.ax.add_patch(self.rect)
         self.artists = [self.rect]
 
-    def ignore(self, event):
-        # docstring inherited
-        return super().ignore(event) or not self.visible
-
     def _press(self, event):
         """Button press event handler."""
-        self.rect.set_visible(self.visible)
-        if self.span_stays:
-            self.stay_rect.set_visible(False)
-            # really force a draw so that the stay rect is not in
-            # the blit background
-            if self.useblit:
-                self.canvas.draw()
-        xdata, ydata = self._get_data(event)
-        if self.direction == 'horizontal':
-            self.pressv = xdata
+        if self.interactive and self.rect.get_visible():
+            self._set_active_handle(event)
         else:
-            self.pressv = ydata
+            self.active_handle = None
 
-        self._set_span_xy(event)
-        return False
+        if self.active_handle is None or not self.interactive:
+            # Clear previous rectangle before drawing new rectangle.
+            self.update()
 
-    def _release(self, event):
-        """Button release event handler."""
-        if self.pressv is None:
-            return
+        if not self.interactive:
+            v = event.xdata if self.direction == 'horizontal' else event.ydata
+            self.extents = v, v
 
-        self.rect.set_visible(False)
+        self.set_visible(self.visible)
 
-        if self.span_stays:
-            self.stay_rect.set_x(self.rect.get_x())
-            self.stay_rect.set_y(self.rect.get_y())
-            self.stay_rect.set_width(self.rect.get_width())
-            self.stay_rect.set_height(self.rect.get_height())
-            self.stay_rect.set_visible(True)
+        return False
 
-        self.canvas.draw_idle()
-        vmin = self.pressv
-        xdata, ydata = self._get_data(event)
-        if self.direction == 'horizontal':
-            vmax = xdata or self.prev[0]
-        else:
-            vmax = ydata or self.prev[1]
+    def _release(self, event):
+        """Button release event handler."""
+        if not self.interactive:
+            self.rect.set_visible(False)
 
-        if vmin > vmax:
-            vmin, vmax = vmax, vmin
+        vmin, vmax = self.extents
         span = vmax - vmin
-        if self.minspan is not None and span < self.minspan:
+        if span < self.minspan:
+            self.set_visible(False)
+            self.update()
             return
+
         self.onselect(vmin, vmax)
-        self.pressv = None
+        self.update()
```

Comment:
This should be fixed (perhaps in the base class implementation) so that one can set extents *before* the widget is ever drawn (e.g. to start with an already drawn SpanSelector).  Currently, setting `extents` before the first `draw` causes `update` to raise `AttributeError: draw_artist can only be used after an initial draw which caches the renderer` (it's probably just a matter of checking whether _cachedRenderer is set and skip the draw in that case.

Reply:
Done